### ISSUE TITLE
Update interoperability documentation

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#10316]
This release is seeing a lot of updates to our interoperability stories.  These updates should be documented, but it may be good to wait on that until closer to the release, since the instructions are going to continue to evolve.

There are aspects of the Cython path of interoperability that can be a little tricky.  Some of that will go away, but others should be explicitly stated to avoid user issues.  I'm going to track what needs to be done here so that I don't forget it when I do get to writing the documentation (instead of continually updating the instructions)

**Comment by bradcray** [comment]
Writing these up sooner rather than later would also be a helpful way of communicating to others on the team about what you're building and getting their feedback.  I suspect that most of what we know from demos, sprint planning, and code reviews is vague and scattered enough that we wouldn't have specific thoughts or feedback.  Reading the documentation for what we have (or what we are building) would be a good way to get feedback on the approach before the release is frozen.

**Comment by lydia-duncan** [comment]
Need to note that Chapel's compiler and linker must match Cython's for ABI compatibility.  Cython seems to use gcc by default (though there are some systems where it may choose something else) and this is determined when the package is installed, so cannot be updated on the fly depending on what CHPL_TARGET_COMPILER is at the time.  Therefore, Chapel must use those settings when interoperating with Python.  On a Cray, though Cython seems to use gcc to install regardless of what the default PrgEnv is, we should definitely use the PrgEnv wrappers in order to support access to installed modules and the like.

To summarize:
On a linux box - use gcc, usually
On OSX - gcc or clang is fine, since gcc is just a wrapper to clang anyways
On a Cray - use PrgEnv-gnu

**Comment by lydia-duncan** [comment]
> Reading the documentation for what we have (or what we are building) would be a good way to get feedback on the approach before the release is frozen.

That's fair.  Would a sort of ad hoc, summary (say, this issue) be sufficient so I don't have to continually wordsmith as things change?

**Comment by bradcray** [comment]
I'm not sure what format would be best / result in the least work, but I definitely think it's fine to start with rougher notes and improve / polish over time.

> Chapel's compiler and linker must match Cython's for ABI compatibility

Ignorant question:  Are, say, gcc and icc generally not ABI compatible?

**Comment by lydia-duncan** [comment]
>     Chapel's compiler and linker must match Cython's for ABI compatibility

> Ignorant question: Are, say, gcc and icc generally not ABI compatible?

According to [this link](https://software.intel.com/en-us/node/522750), the answer is generally yes but there may be some flags that are necessary to do so.  I'm not sure if Cython would provide them under those circumstances.
